reana_server/rest/workflows.py

```diff
@@ -365,57 +363,67 @@ def create_workflow(user):  # noqa
-                    workflow_name,
-                    git_branch,
-                    git_commit_sha,
-                ) = _get_reana_yaml_from_gitlab(request.json, user.id_)
-                git_data = {
-                    "git_url": git_url,
-                    "git_branch": git_branch,
-                    "git_commit_sha": git_commit_sha,
-                }
-            else:
-                # validate against schema
-                git_data = {}
-                reana_spec_file = request.json
-                workflow_name = ""
-            workflow_engine = reana_spec_file["workflow"]["type"]
-        elif request.args.get("spec"):
-            return jsonify("Not implemented"), 501
-        else:
+        json_body_is_not_present = not request.json
+        if json_body_is_not_present:
             raise Exception(
-                "Either remote repository or a reana spec need to \
-            be provided"
+                "Either remote repository or a reana specification needs to be provided"
             )
 
-        if workflow_engine not in REANA_WORKFLOW_ENGINES:
-            raise Exception("Unknown workflow type.")
+        request_from_gitlab = "object_kind" in request.json
+        if request_from_gitlab:
+            (
+                reana_spec_file,
+                git_url,
+                workflow_name,
+                git_branch,
+                git_commit_sha,
+            ) = _get_reana_yaml_from_gitlab(request.json, user.id_)
+            git_data = {
+                "git_url": git_url,
+                "git_branch": git_branch,
+                "git_commit_sha": git_commit_sha,
+            }
+        else:
+            git_data = {}
+            reana_spec_file = request.json
+            workflow_name = request.args.get("workflow_name", "")
 
-        workflow_name = request.args.get("workflow_name", workflow_name)
         validate_workflow_name(workflow_name)
         if is_uuid_v4(workflow_name):
             return jsonify({"message": "Workflow name cannot be a valid UUIDv4."}), 400
+
+        if user.has_exceeded_quota() and request_from_gitlab:
+            message = f"User quota exceeded. Please check {REANA_HOSTNAME}"
+            _fail_gitlab_commit_build_status(user, git_url, git_commit_sha, message)
+            return jsonify({"message": "Gitlab webhook was processed"}), 200
+        elif user.has_exceeded_quota():
+            raise REANAQuotaExceededError()
+
+        workflow_engine = reana_spec_file["workflow"]["type"]
+        if workflow_engine not in REANA_WORKFLOW_ENGINES:
+            raise Exception("Unknown workflow type.")
+
+        operational_options = validate_operational_options(
+            workflow_engine, reana_spec_file.get("inputs", {}).get("options", {})
+        )
+
         workflow_dict = {
             "reana_specification": reana_spec_file,
             "workflow_name": workflow_name,
+            "operational_options": operational_options,
```

Comment:
Hmm, haven't tested the part concerning operational options. Do we have an example?

Reply:
Why do you need to test it? This PR doesn't change how we handle `operational_options`. I just moved it into a separate variable to make dictionary creation nicer.